compaction/task_manager_module.cc

```diff
@@ -254,6 +254,17 @@ future<> run_table_tasks(replica::database& db, std::vector<table_tasks_info> ta
     }
 }
 
+future<tasks::task_manager::task::progress> compaction_task_impl::get_progress(const sstables::compaction_data& cdata, const sstables::compaction_progress_monitor& progress_monitor) const {
+    if (cdata.compaction_size == 0) {
+        co_return get_binary_progress();
+    }
+
+    co_return tasks::task_manager::task::progress{
+        .completed = is_complete() ? cdata.compaction_size : progress_monitor.get_progress(),
+        .total = cdata.compaction_size
```

Comment:
ah, I see that the progress_monitor can be safely reset on the success path.
but here I'd ask if is_complete returns true if scrub compaction fails with an exception and why do we care to reset the progress monitor on the error path?

Reply:
RAII, we construct the generator on compaction's ctor, and destroy it on the dtor. btw, the existence of generator is not needed for get_progress(), after the compaction finishes (either on success or failure).

I am not versed with task manager, but what happens in upper layer if we detect that the task failed? do we call progress or just show some description saying that the task has failed?

And btw, I think we can probably kill is_complete() check above. And I think that's the source of the problem. We're saying task is complete on either success or failure. If task run to completion get_progress() should equal cdata.compaction_size, as completion happens once all readers were exhausted.

Reply:
On failure both progress and error message are shown.

The main reason behind the is_complete check is a situation when the reader skip files, see https://github.com/scylladb/scylladb/pull/14878#issuecomment-1701582519. I believe that for such a situation, when the compaction is finished (but without reading any files) a user should see that it's 100% done